feat: implement file explorer component (closes #201)

benschlegel opened the pull request:
# What

This implements an explorer that gives you an overview of your file structure and lets you navigate it (described in more detail in #201)

<img alt="explorer" src="https://i.imgur.com/87HwzQL.png">

# Features
- navigation to pages via files
- saving/reading of explorer state (what folders are opened/closed)
- supports infinitely nested folders
- lots of config options to customize behavior

# Config

Config options for explorer include:
- `title`: title of explorer component (defaults to `Explorer`)
- `folderClickBehavior`: either `collapse` (collapses folders when clicking on folder label) or `link` (navigate to folder overview page [e.g. `advanced` links to `www.quartz.com/advanced`)
- `folderDefaultState`: either `collapsed` (all folders closed by default) or `open` (all folders are opened by default)
- `useSavedState`: boolean, wether to read/save explorer state (what folders are opened)

# How

The component only has access to `allFiles` as an array (includes all files + path). To make rendering the explorer easier/possible, I first bring all file paths into a tree structure `FileNode` (root path as root node, child paths as child nodes). 

A main `Explorer.tsx` component renders the Explorer outline, while `ExplorerNode.tsx` recursively renders all folders + children and files.

All logic is in `explorer.inline.ts`, responsible for things like collapsing folders, navigating, handling events, saving/reading localStorage.

benschlegel committed:
feat: add basic explorer structure„

benschlegel committed:
feat: integrate new component/plugin

benschlegel committed:
feat: add basic explorer structure

benschlegel committed:
feat: add sort to FileNodes

benschlegel committed:
style: improve style for explorer

benschlegel committed:
refactor: remove unused explorer plugin

benschlegel committed:
refactor: clean explorer structure, fix base (toc)

benschlegel committed:
refactor: clean css, respect displayClass

benschlegel committed:
style: add styling to chevron

benschlegel committed:
refactor: clean up debug statements

benschlegel committed:
refactor: remove unused import

jackyzha0 commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -0,0 +1,117 @@
+    <>
+      {node.file ? (
+        <li key={node.file.slug}>
+          <a href={`/${node.file.slug}`} data-for={node.file.slug}>
```
its incorrect to always go from the root, we need to resolve using relative pathing, see backlinks to see how its done: https://github.com/jackyzha0/quartz/blob/v4/quartz/components/Backlinks.tsx#L15

benschlegel committed:
fix: clicking folder icon sometimes turns invisible

benschlegel committed:
refactor: clean css

benschlegel committed:
feat(explorer): add config for title

benschlegel committed:
feat: add config for folder click behavior

benschlegel committed:
fix: `no-pointer` not being set for all elements

new approach, have one `no-pointer` class, that removes pointer events and one `clickable` class on the svg and button (everything that can normally be clicked). then, find all children with `clickable` and toggle `no-pointer`

benschlegel committed:
fix: bug where nested folders got incorrect height

this fixes the bug where nested folders weren't calculating their total height correctly. done by adding class to main container of all children and calculating total

benschlegel committed:
feat: introduce `folderDefaultState` config

benschlegel committed:
feat: store depth for explorer nodes

benschlegel committed:
feat: implement option for collapsed state + bug fixes

folderBehavior: "link" still has bad styling, but major bugs with pointers fixed (not clean yet, but working)

benschlegel committed:
fix: default folder icon rotation

benschlegel committed:
fix: hitbox problem with folder links, fix style

benschlegel committed:
fix: redirect url for nested folders

benschlegel committed:
fix: inconsistent behavior with 'collapseFolders' opt

benschlegel committed:
chore: add comments to `ExplorerNode`

benschlegel committed:
feat: save explorer state to local storage (not clean)

benschlegel committed:
feat: rework `getFolders()`, fix localstorage read + write

benschlegel committed:
feat: set folder state from localStorage

needs serious refactoring but functional (except folder icon orientation)

benschlegel committed:
fix: folder icon orientation after local storage

benschlegel committed:
feat: add config for `useSavedState`

benschlegel committed:
refactor: clean `explorer.inline.ts`

remove unused functions, comments, unused code, add types to EventHandler

benschlegel committed:
refactor: clean explorer

merge `isSvg` paths, remove console logs

benschlegel committed:
refactor: add documentation, remove unused funcs

benschlegel commented:
ready for review now @jackyzha0, updated the PR description and fixed all major bugs (from my testing)

jackyzha0 commented on quartz.layout.ts:
```diff
@@ -22,6 +22,7 @@ export const defaultContentPageLayout: PageLayout = {
     Component.Search(),
     Component.Darkmode(),
     Component.DesktopOnly(Component.TableOfContents()),
+    Component.DesktopOnly(Component.Explorer()),
```
I think we should move table of contents to the right sidebar between graph and backlinks now that explorer exists :)

benschlegel replied:
done in 552cc374b6ae820e6c25772d806d2f02808e9263

jackyzha0 commented on quartz/components/Explorer.tsx:
```diff
@@ -0,0 +1,71 @@
+import { GlobalConfiguration } from "../cfg"
+
+// Options interface defined in `ExplorerNode` to avoid circular dependency
+const defaultOptions = (cfg: GlobalConfiguration): Options => ({
```
`cfg` is unused here, intentional?

benschlegel replied:
no, removed in c3c6bad7a0930657bf42da7c55632706c5c9a575

jackyzha0 commented on quartz/styles/base.scss:
```diff
@@ -446,7 +446,10 @@ video {
+  max-height: 300;
+  display: flex;
+  flex-direction: column;
+  flex-shrink: 0;
```
Works without these 3, would prefer not to add flex where we don't need it

benschlegel replied:
fixed in 2a2ee25ee3b3777b68c0eade50f15f6983bd0c86

jackyzha0 commented on quartz/components/styles/explorer.scss:
```diff
@@ -0,0 +1,131 @@
+  }
+}
+
+.folder-icon {
```
```suggestion
.folder-icon {
  margin-left: -5px;
```

benschlegel replied:
solved this by improving the svg viewbox (no more apperent margin) and adding margin-right instead of negative left margin (in 2a2ee25ee3b3777b68c0eade50f15f6983bd0c86)

jackyzha0 commented on quartz/components/styles/explorer.scss:
```diff
@@ -0,0 +1,131 @@
+
+svg > polyline {
+  pointer-events: none;
+}
```
```suggestion
svg {
  pointer-events: all;
  & > polyline {
    pointer-events: none;
  }
}
```

benschlegel replied:
87b15e6bc16c9b847d41ead1e16b7e4e35ce13f7

jackyzha0 commented on quartz/components/styles/explorer.scss:
```diff
@@ -0,0 +1,131 @@
+
+ul.overflow::after {
+  background: none;
+}
```
```suggestion
```

jackyzha0 replied:
the background on the overflow helps to indicate more content, we should keep this. if you think the 50px padding is too big, we can adjust it down to 30px (`ol.overflow` in `base.scss`)

benschlegel replied:
fixed in (aec2ebc1568f8148d8808263075cbd95139e290f)

personally, i dont like the fade effect thats there without `background: none` as seen here:

<img width="211" alt="image" src="https://github.com/jackyzha0/quartz/assets/31989404/2f43fa87-c2cb-4438-807d-b20db9485dc3">

it seems inconsistent, as a really long TOC does not have the same effect:

<img width="478" alt="image" src="https://github.com/jackyzha0/quartz/assets/31989404/7c14c75e-40bb-4251-8890-b0c9d6789311">

jackyzha0 replied:
oh hm that's weird, the TOC should have the effect, that's a bug

jackyzha0 replied:
i'd like some sort of affordance that there's more stuff if you scroll so as long as we have that i'm happy

benschlegel replied:
okay yeah made sense, my initial confusion came from the fact that it wasnt consistent with TOC behavior, so i tried to manually remove it, but thats a bug fix for a seperate PR then to fix TOC/overflow class

benschlegel commented:
found a different solution that uses flexbox (switches between flex 1 and 0) instead of calculating and setting maxHeight, gonna try to rewrite explorer to use that instead before merging (should be a lot more stable and cleaner)

benschlegel committed:
feat: rework folder collapse logic

use grids instead of jank scuffed solution with calculating total heights

jackyzha0 commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  private insert(file: DataWrapper, depth: number) {
+    this.depth = ++depth
```
why do we `++depth` here?

benschlegel replied:
it technically worked, when printing the tree, all nodes had the correct depth:

<img width="327" alt="image" src="https://github.com/jackyzha0/quartz/assets/31989404/75c0e2c2-75be-4fa3-ab8a-90e630572497">

cleaned and made this better readable by removing depth arg and calculating it directly (changed in ad94422235645fa964945b25fc0e9a3192e1c6c8)

jackyzha0 commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -0,0 +1,205 @@
+        // Single file node
+        <li key={node.file.slug} class="no-pointer">
+          <a
+            href={`/${node.file.slug}`}
```
incorrect to go from `/` as this will break for any Quartz hosted on a subpath (e.g. username.github.io/quartz)

benschlegel replied:
should I use the `baseUrl` from `cfg` instead and add paths after based on that instead?

jackyzha0 replied:
that would work but let's not be dependent on `baseUrl`. you should do `resolveRelative`: https://github.com/jackyzha0/quartz/blob/v4/quartz/components/Backlinks.tsx#L15

benschlegel replied:
done in 2acdf14570a0ba6a2460ed4ca7bd76012057437d

jackyzha0 commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -0,0 +1,205 @@
+          {node.name !== "" && (
+            // Node with entire folder
+            // Render svg button + folder name, then children
+            <div class={`folder-container ${isInvisInner ? "no-pointer" : ""}`}>
```
i wonder if we should just make a component for a collapsable list lol

benschlegel replied:
yes, it would make life so much easier, we should probably do that after merging this in a seperate PR. I found a really clean way to do this with a grid based approach (example [here](https://www.stefanjudis.com/snippets/how-to-animate-height-with-css-grid/)) and am currently working on refactoring the file explorer to use this approach. saves so much jank with setting pointer-events: none, makes life easier and we should use it everywhere (no more setting heights manually, etc)

jackyzha0 replied:
that would be super epic, excited to see what you come up with!

benschlegel replied:
fixed this for explorer for now, could do this for the rest of quartz in the future in a separate issue/PR

jackyzha0 commented:
I think we should try to use the title of the folder where possible (e.g. if a folder `a` has a file `a/index.md` we should use the title of that file. So the `feature` folder would be `Feature List`.

We should also add the ability for users to pass in a map function and a filter function so people can transform the display name for folders and notes and also hide unwanted ones (e.g. remove the `tags` folder)

benschlegel commented:
offtopic but the grid based approach re-write also fixes the "whitespace at the bottom" bug. gonna first rewrite the explorer to use cleaner collapsables, then implement your feedback :)

benschlegel committed:
refactor: remove depth arg from insert

benschlegel committed:
feat: restore collapse functionality to clicks

allow folder icon + folder label to collapse folders again

benschlegel committed:
refactor: remove `pointer-event` jank

benschlegel committed:
feat: improve svg viewbox + remove unused props

benschlegel committed:
feat: use css selector to toggle icon

rework folder icon to work purely with css instead of JS manipulation

benschlegel committed:
refactor: remove unused cfg

benschlegel committed:
feat: move TOC to right sidebar

benschlegel committed:
refactor: clean css

benschlegel committed:
style: fix overflow + overflow margin

benschlegel committed:
fix: use `resolveRelative` to resolve file paths

benschlegel committed:
fix: `defaultFolderState` config option

benschlegel commented:
> I think we should try to use the title of the folder where possible (e.g. if a folder `a` has a file `a/index.md` we should use the title of that file. So the `feature` folder would be `Feature List`.
> 
> We should also add the ability for users to pass in a map function and a filter function so people can transform the display name for folders and notes and also hide unwanted ones (e.g. remove the `tags` folder)

those sound like good ideas and ill gladly work on them in the future/next few days, but i think we should first merge a "base state" of the explorer, that should already be working for basic use cases and the expand it afterwards in a separate PR to improve it with the features you mentioned. otherwise, people will have to be stuck waiting even longer until all the new enhancements are working as expected, the mapping might be a bit more complex and both mapping/filtering would take a lot of testing to make sure it works as intended.

the first part seems straight forward, but could you outline how you expect the maps/filter functions to behave in an issue or something? how do you expect the syntax for mappers/filters to look like/how would you like them to get passed? (as functions via options or some other way)

benschlegel commented:
I now implemented all your feedback and switched out the collapsables internally (from annoying jank way with setting heights/pointer-events manually to new grid based approached that does all of that automatically and only requires the `open` class to be set to work).

Also made folder icons work **only** through css selectors, no more state set through JS required, automatically updates if `open` state changes. Fixed a few small bugs ive noticed while refactoring.

Is now open to be reviewed again, should be ready now.

jackyzha0 commented:
> > I think we should try to use the title of the folder where possible (e.g. if a folder `a` has a file `a/index.md` we should use the title of that file. So the `feature` folder would be `Feature List`.
> > We should also add the ability for users to pass in a map function and a filter function so people can transform the display name for folders and notes and also hide unwanted ones (e.g. remove the `tags` folder)
> 
> those sound like good ideas and ill gladly work on them in the future/next few days, but i think we should first merge a "base state" of the explorer, that should already be working for basic use cases and the expand it afterwards in a separate PR to improve it with the features you mentioned. otherwise, people will have to be stuck waiting even longer until all the new enhancements are working as expected, the mapping might be a bit more complex and both mapping/filtering would take a lot of testing to make sure it works as intended.
> 
> the first part seems straight forward, but could you outline how you expect the maps/filter functions to behave in an issue or something? how do you expect the syntax for mappers/filters to look like/how would you like them to get passed? (as functions via options or some other way)

Sounds good, yup this is not blocking for me. I think passing in functions as props is fine for those. Map would take `FileNode => FileNode`, filter would take `FileNode => bool`

jackyzha0 commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -0,0 +1,197 @@
+// @ts-ignore
+import { Data } from "vfile"
```
prefer `QuartzPluginData` to `Data` just for semantics

jackyzha0 replied:
wait I meant that we should use `QuartzPluginData` instead of `Data` haha lol

benschlegel replied:
ahh my bad, was late already 😅
fixed now

jackyzha0 commented on quartz/components/scripts/explorer.inline.ts:
```diff
@@ -0,0 +1,139 @@
+    explorerState = JSON.parse(storageTree)
+    explorerState.map((folder) => {
+      // grab <li> element for matching folder path
+      const folderLI = document.querySelector(`[data-folderpath='/${folder.path}']`) as HTMLElement
```
nit: should be named `folderLi` and `folderUl` respectively

jackyzha0 reviewed (approved):
lgtm! final request: can you write a documentation page for the new component?

benschlegel committed:
refactor: rename import, rename `folderLi` + ul

benschlegel commented:
> lgtm! final request: can you write a documentation page for the new component?

Yeah, was planning to do that anyways, but ill get to that tomorrow as its 1am here already :)

Also fixed the last 2 issues

benschlegel committed:
fix: use `QuartzPluginData` type

benschlegel committed:
docs: add explorer documentation

benschlegel commented:
Added documentation for this component now (open to feedback)
